src/backend/distributed/utils/connection_management.c

```diff
@@ -0,0 +1,541 @@
+/*-------------------------------------------------------------------------
+ *
+ * connection_management.c
+ *   Central management of connections and their life-cycle
+ *
+ * Copyright (c) 2016, Citus Data, Inc.
+ *
+ *-------------------------------------------------------------------------
+ */
+
+#include "postgres.h"
+
+#ifdef HAVE_POLL_H
+#include <poll.h>
+#endif
+
+#include "libpq-fe.h"
+
+#include "miscadmin.h"
+
+#include "access/hash.h"
+#include "commands/dbcommands.h"
+#include "distributed/connection_management.h"
+#include "distributed/metadata_cache.h"
+#include "distributed/hash_helpers.h"
+#include "mb/pg_wchar.h"
+#include "utils/hsearch.h"
+#include "utils/memutils.h"
+
+
+HTAB *ConnectionHash = NULL;
+MemoryContext ConnectionContext = NULL;
+
+static uint32 ConnectionHashHash(const void *key, Size keysize);
+static int ConnectionHashCompare(const void *a, const void *b, Size keysize);
+static MultiConnection * StartConnectionEstablishment(ConnectionHashKey *key);
+
+
+/*
+ * Initialize per-backend connection management infrastructure.
+ */
+void
+InitializeConnectionManagement(void)
+{
+	HASHCTL info;
+	uint32 hashFlags = 0;
+
+
+	/*
+	 * Create a single context for connection and transaction related memory
+	 * management. Doing so, instead of allocating in TopMemoryContext, makes
+	 * it easier to associate used memory.
+	 */
+	ConnectionContext = AllocSetContextCreate(TopMemoryContext, "Connection Context",
+											  ALLOCSET_DEFAULT_MINSIZE,
+											  ALLOCSET_DEFAULT_INITSIZE,
+											  ALLOCSET_DEFAULT_MAXSIZE);
+
+	/* create (host,port,user,database) -> [connection] hash */
+	memset(&info, 0, sizeof(info));
+	info.keysize = sizeof(ConnectionHashKey);
+	info.entrysize = sizeof(ConnectionHashEntry);
+	info.hash = ConnectionHashHash;
+	info.match = ConnectionHashCompare;
+	info.hcxt = ConnectionContext;
+	hashFlags = (HASH_ELEM | HASH_FUNCTION | HASH_CONTEXT | HASH_COMPARE);
+
+	ConnectionHash = hash_create("citus connection cache (host,port,user,database)",
+								 64, &info, hashFlags);
+}
+
+
+/*
+ * Perform connection management activity after the end of a transaction. Both
+ * COMMIT and ABORT paths are handled here.
+ *
+ * This is called by Citus' global transaction callback.
+ */
+void
+AtEOXact_Connections(bool isCommit)
+{
+	HASH_SEQ_STATUS status;
+	ConnectionHashEntry *entry;
+
+	/*
+	 * Close all remote connections if necessary anymore (i.e. not session
+	 * lifetime), or if in a failed state.
+	 */
+	hash_seq_init(&status, ConnectionHash);
+	while ((entry = (ConnectionHashEntry *) hash_seq_search(&status)) != 0)
+	{
+		ListCell *previousCell = NULL;
+		ListCell *nextCell = NULL;
+		ListCell *connectionCell = NULL;
+
+		/*
+		 * Have to iterate "manually", to be able to delete connections in the
+		 * middle of the list.
+		 */
+		for (connectionCell = list_head(entry->connections);
```

Comment:
I did, but it doesn't really seem like a beneficial change, especially due to the "forced" placement of static functions... :(